src/nix-env/nix-env.cc

```diff
@@ -224,6 +224,100 @@ static void checkSelectorUse(DrvNames & selectors)
 }
 
 
+namespace {
+
+template <class Container, class F>
+void forEachWithIndex(Container & container, F && func) {
+    std::size_t index = 0;
+    for (auto & item : container) {
+        func(item, index);
+        ++index;
+    }
+}
+
+std::set<std::string> fuzzySearch(const DrvInfos & allElems, const DrvName & name) {
+    constexpr std::size_t maxResults = 3;
+    std::set<std::string> result;
+    for (const auto & drvInfo : allElems) {
+        const auto drvName = DrvName { drvInfo.queryName() };
+        if (hasPrefix(drvName.name, name.name)) {
+            result.emplace(drvName.name);
+
+            if (result.size() >= maxResults) {
+                break;
+            }
+        }
+    }
+    return result;
+}
+
+struct Match
+{
+    DrvInfo drvInfo;
+    std::size_t index;
+
+    Match(DrvInfo drvInfo_, std::size_t index_)
```

Comment:
IMHO we don't need trivial constructors like this in C++20 anymore since we can use designated initializers, i.e.
```
Match { .drvInfo = ..., .index = ... }
```
which are easier to read.

Reply:
True but in this case I was aiming to enable the use of `emplace`

Reply:
Not sure if that really matters. But for efficiency you probably want to declare `DrvInfo && drvInfo` because otherwise it's passed by value which is probably not what you want.

Reply:
This is actually the preferred way to pass for sink-like semantics, it is optimized very well. If I use && then only rvalues would be accepted, which is an unnecessary restriction for this usecase.

https://godbolt.org/z/KvsdWh13d
Flip 1 on line 20 to make sure the generated code is the same